File: hooks/edit/tests/rust_edit_test.py
# pyright: reportUninitializedInstanceVariable=false
# setUp fills these per test, which is where a unittest fixture belongs. The rule asks for a
# class-body assignment or __init__ instead, neither of which can hold a per-test temp dir.
# The class-body annotations still carry the types.
"""Integration tests for hooks/edit/rust_pre_edit.py and rust_post_edit.py.

cargo is replaced by a stub on PATH: the assertions read what the hooks do with clippy's
output rather than compiling a real crate.

Run: python3 hooks/edit/tests/rust_edit_test.py
"""

import json
import os
import shutil
import subprocess
import sys
import tempfile
import unittest
from pathlib import Path
from typing import override

HOOK_DIR = Path(__file__).resolve().parents[1]

sys.path.insert(0, str(Path(__file__).resolve().parents[2] / "_lib"))

import hook_harness  # noqa: E402

PRE = HOOK_DIR / "rust_pre_edit.py"
POST = HOOK_DIR / "rust_post_edit.py"

# Records the subcommand it was called with, and prints findings for two files so the
# ordering assertion has something to reorder.
STUB_CARGO = """#!/bin/sh
echo "$1" >> "$CARGO_CALLS"
[ "$1" = "clippy" ] || exit 0
[ -n "$CARGO_SILENT" ] && exit 0
echo "src/other.rs:1:1: warning: unused variable"
echo "src/other.rs:2:1: warning: needless return"
echo "src/lib.rs:9:1: warning: this looks like the edited file"
exit 0
"""


class TestRustEdit(unittest.TestCase):
    tmpdir: tempfile.TemporaryDirectory[str]
    repo: Path
    calls: Path
    env: dict[str, str]

    @override
    def setUp(self) -> None:
        self.tmpdir = tempfile.TemporaryDirectory(prefix="rust-edit-tests-")
        self.addCleanup(self.tmpdir.cleanup)
        root = Path(self.tmpdir.name)

        self.repo = root / "repo"
        (self.repo / "src").mkdir(parents=True)
        _ = subprocess.run(["git", "-C", str(self.repo), "init", "-q"], check=True)
        hook_harness.disable_background_git_maintenance(self.repo)
        (self.repo / "src" / "lib.rs").touch()
        (self.repo / "src" / "other.rs").touch()

        stub_bin = root / "bin"
        stub_bin.mkdir()
        stub = stub_bin / "cargo"
        _ = stub.write_text(STUB_CARGO, encoding="utf-8")
        stub.chmod(0o755)

        self.calls = root / "calls"
        self.calls.touch()
        self.env = {
            **os.environ,
            "PATH": f"{stub_bin}{os.pathsep}{os.environ['PATH']}",
            "CARGO_CALLS": str(self.calls),
        }
        # Every assertion below that expects silence also passes when cargo is unreachable,
        # which is what a replaced (rather than merged) env produces.
        self.assertEqual(shutil.which("cargo", path=self.env["PATH"]), str(stub))

    def run_hook(self, hook: Path, path: Path | str, **env: str) -> str:
        payload = {"tool_name": "Edit", "tool_input": {"file_path": str(path)}}
        return hook_harness.run(hook, payload, {**self.env, **env})

    def cargo_calls(self) -> list[str]:
        return self.calls.read_text(encoding="utf-8").split()

    def test_the_fixture_repository_answers_zero_for_gc_auto(self) -> None:
        """T-421 the python tests' fixture repository answers 0 for gc.auto, read back through
        git config"""
        result = subprocess.run(
            ["git", "-C", str(self.repo), "config", "gc.auto"],
            check=False,
            capture_output=True,
            text=True,
        )
        gc_auto = result.stdout.strip() if result.returncode == 0 else ""
        self.assertEqual(gc_auto, "0")

    def test_a_non_rust_edit_never_starts_cargo(self) -> None:
        """T-001: An edit to anything but .rs does not launch cargo"""
        self.assertEqual(self.run_hook(PRE, self.repo / "README.md"), "")
        self.assertEqual(self.cargo_calls(), [])

    def test_a_rust_file_outside_a_repo_is_skipped(self) -> None:
        """T-002: A .rs outside git control is out of scope"""
        loose = Path(self.tmpdir.name) / "loose.rs"
        loose.touch()
        self.assertEqual(self.run_hook(PRE, loose), "")
        self.assertEqual(self.cargo_calls(), [])

    def test_the_edited_file_findings_come_first(self) -> None:
        """T-003: Findings for the edited file lead the list"""
        # clippy reads the whole workspace, so trimming can drop the file that was edited.
        out = self.run_hook(PRE, self.repo / "src" / "lib.rs")
        context = json.loads(out)["hookSpecificOutput"]["additionalContext"]
        self.assertIn("src/lib.rs", context.splitlines()[0])

    def test_pre_edit_declares_its_event(self) -> None:
        """T-004: It returns as a PreToolUse"""
        out = self.run_hook(PRE, self.repo / "src" / "lib.rs")
        self.assertEqual(json.loads(out)["hookSpecificOutput"]["hookEventName"], "PreToolUse")

    def test_post_edit_formats_then_lints(self) -> None:
        """T-005: After an edit, fmt runs and then clippy returns"""
        out = self.run_hook(POST, self.repo / "src" / "lib.rs")
        self.assertEqual(self.cargo_calls(), ["fmt", "clippy"])
        self.assertEqual(json.loads(out)["hookSpecificOutput"]["hookEventName"], "PostToolUse")

    def test_a_clean_clippy_says_nothing(self) -> None:
        """T-006: With no finding it returns nothing"""
        # Injecting an empty additionalContext hands the reader nothing.
        out = self.run_hook(PRE, self.repo / "src" / "lib.rs", CARGO_SILENT="1")
        self.assertEqual(out, "")


if __name__ == "__main__":
    unittest.main(verbosity=2)

File: hooks/edit/tests/rust-edit.test.ts
/// <reference types="node" />
// Unit U-003: hooks/edit/rust_pre_edit.ts and hooks/edit/rust_post_edit.ts both compose
// hooks/_lib/rust_target.ts's target/fmt/clippyOutput, which is why the plan folds
// hooks/edit/tests/rust_edit_test.py's two-hook coverage into this one test file. cargo is
// replaced by a stub on PATH exactly as rust_edit_test.py's STUB_CARGO and
// hooks/_lib/tests/rust-target.test.ts's installStubCargo do: the assertions read what the
// hooks do with the stub's clippy output, not a real crate.
import assert from "node:assert/strict";
import { execFileSync } from "node:child_process";
import { chmodSync, mkdirSync, mkdtempSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import path from "node:path";
import test from "node:test";
import { fileURLToPath } from "node:url";
import { run } from "../../_lib/tests/_hook-harness.ts";

const HERE = path.dirname(fileURLToPath(import.meta.url));
const PRE = path.join(HERE, "..", "rust_pre_edit.ts");
const POST = path.join(HERE, "..", "rust_post_edit.ts");

// Records the subcommand it was called with, and (unless CARGO_SILENT) prints one finding
// naming the edited file -- the smallest stub that lets both the "reports" and the "stays
// silent when there are none" halves of each scenario run. Mirrors rust_edit_test.py's
// STUB_CARGO.
const STUB_CARGO = `#!/bin/sh
echo "$1" >> "$CARGO_CALLS"
[ "$1" = "clippy" ] || exit 0
[ -n "$CARGO_SILENT" ] && exit 0
echo "src/lib.rs:9:1: warning: this looks like the edited file"
exit 0
`;

interface Fixture {
  repo: string;
  env: NodeJS.ProcessEnv;
}

function fixture(): Fixture {
  const root = mkdtempSync(path.join(tmpdir(), "rust-edit-tests-"));
  const repo = path.join(root, "repo");
  mkdirSync(path.join(repo, "src"), { recursive: true });
  execFileSync("git", ["-C", repo, "init", "-q"], { stdio: "ignore" });
  writeFileSync(path.join(repo, "src", "lib.rs"), "");
  writeFileSync(path.join(repo, "README.md"), "");

  const binDir = path.join(root, "bin");
  mkdirSync(binDir);
  const stub = path.join(binDir, "cargo");
  writeFileSync(stub, STUB_CARGO);
  chmodSync(stub, 0o755);

  const calls = path.join(root, "calls");
  writeFileSync(calls, "");

  return {
    repo,
    env: {
      ...process.env,
      PATH: `${binDir}${path.delimiter}${process.env.PATH}`,
      CARGO_CALLS: calls,
    },
  };
}

function runHook(
  hook: string,
  f: Fixture,
  filePath: string,
  extraEnv: NodeJS.ProcessEnv = {},
): string {
  const payload = { tool_name: "Edit", tool_input: { file_path: filePath } };
  return run(hook, payload, { ...f.env, ...extraEnv });
}

test("T-381 the pre-edit hook reports the target it resolves for a .rs path and stays silent for another extension", () => {
  const f = fixture();

  const rustOut = runHook(PRE, f, path.join(f.repo, "src", "lib.rs"));
  assert.ok(rustOut, "a .rs edit must report the target it resolved, but the hook stayed silent");
  const parsed = JSON.parse(rustOut) as {
    hookSpecificOutput: { hookEventName: string; additionalContext: string };
  };
  assert.equal(parsed.hookSpecificOutput.hookEventName, "PreToolUse");
  assert.match(
    parsed.hookSpecificOutput.additionalContext,
    /src\/lib\.rs/,
    "the resolved target's own finding must reach the report",
  );

  const otherOut = runHook(PRE, f, path.join(f.repo, "README.md"));
  assert.equal(otherOut, "", "an edit to a non-.rs path must stay silent");
});

test("T-382 the post-edit hook reports clippy findings for an edited .rs file and stays silent when there are none", () => {
  const f = fixture();

  const withFindings = runHook(POST, f, path.join(f.repo, "src", "lib.rs"));
  assert.ok(
    withFindings,
    "an edited .rs file with clippy findings must be reported, but the hook stayed silent",
  );
  const parsed = JSON.parse(withFindings) as {
    hookSpecificOutput: { hookEventName: string; additionalContext: string };
  };
  assert.equal(parsed.hookSpecificOutput.hookEventName, "PostToolUse");
  assert.match(parsed.hookSpecificOutput.additionalContext, /src\/lib\.rs/);

  const clean = runHook(POST, f, path.join(f.repo, "src", "lib.rs"), { CARGO_SILENT: "1" });
  assert.equal(clean, "", "a clean clippy run must stay silent");
});
